<table style="border: none" align="left">
   <tr style="border: none">
      <th style="border: none"><font face="verdana" size="5" color="black"><b>Predict outdoor equipment purchase with IBM Watson Machine Learning</b></th>
      <th style="border: none"><img src="https://github.com/pmservice/customer-satisfaction-prediction/blob/master/app/static/images/ml_icon_gray.png?raw=true" alt="Watson Machine Learning icon" height="40" width="40"></th>
   </tr>
   <tr style="border: none">
       <th style="border: none"><img src="https://github.com/pmservice/wml-sample-models/blob/master/spark/product-line-prediction/images/products_graphics.png?raw=true" alt="Icon" width="800"> </th>
   </tr>
</table>

This notebook contains steps and code to get data from the IBM Data Science Experience Community, create a predictive model, and start scoring new data. This notebook introduces commands for getting data and for basic data cleaning and exploration, pipeline creation, model training, model persistance to Watson Machine Learning repository, model deployment, and scoring.

Some familiarity with Scala is helpful. This notebook uses Scala 2.11 and Apache® Spark 2.0.

You will use a publicly available data set, **GoSales Transactions for Naive Bayes Model**, which details anonymous outdoor equipment purchases. Use the details of this data set to predict clients' interests in terms of product line, such as golf accessories, camping equipment, and others.

## Learning goals

The learning goals of this notebook are:

-  Load a CSV file into an Apache® Spark DataFrame.
-  Explore data.
-  Prepare data for training and evaluation.
-  Create an Apache® Spark machine learning model.
-  Train and evaluate a model.
-  Persist a model in a Watson Machine Learning repository.
-  Deploy a model for online scoring using Wastson Machine Learning API.
-  Score sample scoring data using the Watson Machine Learning API.


## Contents

This notebook contains the following parts:

1.	[Setup](#setup)
2.	[Load and explore data](#load)
3.	[Create spark ml model](#model)
4.	[Persist model](#persistence)
5.	[Predict locally and visualize](#visualization)
6.	[Deploy and score in a Cloud](#scoring)
7.	[Summary and next steps](#summary)

<a id="setup"></a>
## 1. Setup

Before you use the sample code in this notebook, you must perform the following setup tasks:

-  Create a [Watson Machine Learning Service](https://console.ng.bluemix.net/catalog/services/ibm-watson-machine-learning/) instance (a free plan is offered). 
-  Upload **GoSales Transactions** data as a data asset in IBM Data Science Experience.
-  Make sure that you are using a Spark 2.0 kernel.

### Create the GoSales Transactions Data Asset  

The GOSales data is a freely available data set on the Data Science Experience home page.

1.  Go to the [GoSales Transactions for Naive Bayes Model](https://apsportal.ibm.com/exchange-api/v1/entries/8044492073eb964f46597b4be06ff5ea/data?accessKey=9561295fa407698694b1e254d0099600) data card on the Data Science Experience **Community** page and open the card by double-clicking it.
2.  Click the link icon.
4.  Select the link, copy it by pressing Ctrl+C, and then, click **Close**.
5.  In the following cell, replace the **link_to_data** variable value with the link.

<a id="load"></a>
## 2. Load and explore data

In this section you will load the data as an Apache® Spark DataFrame and perform a basic exploration.

Load the data to the Spark DataFrame by using *wget* to upload the data to gpfs and then *read* method. 

**Example**: First, you need to install required packages. You can do it by running the following code. Run it only one time.

In [ ]:
import scala.sys.process._

"wget https://apsportal.ibm.com/exchange-api/v1/entries/8044492073eb964f46597b4be06ff5ea/data?accessKey=9561295fa407698694b1e254d0099600".!

In [4]:
val filename = "data?accessKey=9561295fa407698694b1e254d0099600"

In [13]:
val df_data = spark.
    read.format("org.apache.spark.sql.execution.datasources.csv.CSVFileFormat").
    option("header", "true").
    option("inferSchema", "true").
    load(filename)
df_data.take(5)

Array([Personal Accessories,M,27,Single,Professional], [Personal Accessories,F,39,Married,Other], [Mountaineering Equipment,F,39,Married,Other], [Personal Accessories,F,56,Unspecified,Hospitality], [Golf Equipment,M,45,Married,Retired])

Explore the loaded data by using the following Apache® Spark DataFrame methods:
-  print schema
-  print top ten records
-  count all records

In [14]:
df_data.printSchema()

root
 |-- PRODUCT_LINE: string (nullable = true)
 |-- GENDER: string (nullable = true)
 |-- AGE: integer (nullable = true)
 |-- MARITAL_STATUS: string (nullable = true)
 |-- PROFESSION: string (nullable = true)



As you can see, the data contains five fields. PRODUCT_LINE field is the one we would like to predict (label).

In [15]:
df_data.show()

+--------------------+------+---+--------------+------------+
|        PRODUCT_LINE|GENDER|AGE|MARITAL_STATUS|  PROFESSION|
+--------------------+------+---+--------------+------------+
|Personal Accessories|     M| 27|        Single|Professional|
|Personal Accessories|     F| 39|       Married|       Other|
|Mountaineering Eq...|     F| 39|       Married|       Other|
|Personal Accessories|     F| 56|   Unspecified| Hospitality|
|      Golf Equipment|     M| 45|       Married|     Retired|
|      Golf Equipment|     M| 45|       Married|     Retired|
|   Camping Equipment|     F| 39|       Married|       Other|
|   Camping Equipment|     F| 49|       Married|       Other|
|  Outdoor Protection|     F| 49|       Married|       Other|
|      Golf Equipment|     M| 47|       Married|     Retired|
|      Golf Equipment|     M| 47|       Married|     Retired|
|Mountaineering Eq...|     M| 21|        Single|      Retail|
|Personal Accessories|     F| 66|       Married|       Other|
|   Camp

In [16]:
df_data.count()

60252

As you can see, the data set contains 60252 records.

<a id="model"></a>
## 3. Create an Apache® Spark machine learning model

In this section you will learn how to prepare data, create and train an Apache® Spark machine learning model.

### 3.1: Prepare data

In this subsection you will split your data into: train, test and predict data sets.

In [28]:
val splits = df_data.randomSplit(Array(0.8, 0.18, 0.02), seed = 24L)
val training_data = splits(0).cache()
val test_data = splits(1)
val prediction_data = splits(2)

println("Number of training records: ", training_data.count())
println("Number of testing records : ", test_data.count())
println("Number of prediction records : ", prediction_data.count())

(Number of training records: ,48176)
(Number of testing records : ,10860)
(Number of prediction records : ,1216)


As you can see our data has been successfully split into three data sets: 

-  The train data set, which is the largest group, is used for training.
-  The test data set will be used for model evaluation and is used to test the assumptions of the model.
-  The predict data set will be used for prediction.

### 3.2: Create pipeline and train a model

In this section you will create an Apache® Spark machine learning pipeline and then train the model.

In the first step you need to import the Apache® Spark machine learning packages that will be needed in the subsequent steps.

In [29]:
// Spark ML libraries

import org.apache.spark.ml.classification.RandomForestClassifier
import org.apache.spark.ml.feature.{OneHotEncoder, StringIndexer, IndexToString, VectorAssembler}
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator
import org.apache.spark.ml.{Model, Pipeline, PipelineStage, PipelineModel}
import org.apache.spark.sql.SparkSession

In the following step, convert all the string fields to numeric ones by using the StringIndexer transformer.

In [44]:
val stringIndexer_label = new StringIndexer().setInputCol("PRODUCT_LINE").setOutputCol("label").fit(df_data)
val stringIndexer_prof = new StringIndexer().setInputCol("PROFESSION").setOutputCol("PROFESSION_IX")
val stringIndexer_gend = new StringIndexer().setInputCol("GENDER").setOutputCol("GENDER_IX")
val stringIndexer_mar = new StringIndexer().setInputCol("MARITAL_STATUS").setOutputCol("MARITAL_STATUS_IX")

In the following step, create a feature vector by combining all features together.

In [39]:
val vectorAssembler_features = new VectorAssembler().setInputCols(Array("GENDER_IX", "AGE", "MARITAL_STATUS_IX", "PROFESSION_IX")).setOutputCol("features")

Next, define estimators you want to use for classification. Random Forest is used in the following example.

In [41]:
val rf = new RandomForestClassifier().setLabelCol("label").setFeaturesCol("features").setNumTrees(10)

Finally, indexed labels back to original labels.

In [45]:
val labelConverter = new IndexToString().setInputCol("prediction").setOutputCol("predictedLabel").setLabels(stringIndexer_label.labels)

Let's build the pipeline now. A pipeline consists of transformers and an estimator.

In [47]:
val pipeline_rf = new Pipeline().setStages(Array(stringIndexer_label, stringIndexer_prof, stringIndexer_gend, stringIndexer_mar, vectorAssembler_features, rf, labelConverter))

Now, you can train your Random Forest model by using the previously defined **pipeline** and **training data**.

In [49]:
training_data.printSchema()

root
 |-- PRODUCT_LINE: string (nullable = true)
 |-- GENDER: string (nullable = true)
 |-- AGE: integer (nullable = true)
 |-- MARITAL_STATUS: string (nullable = true)
 |-- PROFESSION: string (nullable = true)



In [51]:
val model_rf = pipeline_rf.fit(training_data)

You can check your **model accuracy** now. To evaluate the model, use **test data**.

In [58]:
val predictions = model_rf.transform(test_data)
val evaluatorRF = new MulticlassClassificationEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("accuracy")
val accuracy = evaluatorRF.evaluate(predictions)
println("Accuracy = " + accuracy)
println("Test Error = " + (1.0 - accuracy))

Accuracy = 0.5736648250460405
Test Error = 0.42633517495395945


You can tune your model now to achieve better accuracy. For simplicity of this example tuning section is omitted.

<a id="persistence"></a>
## 4. Persist model

In this section you will learn how to store your pipeline and model in Watson Machine Learning repository by using Scala client libraries.

First, you must import client libraries.

In [13]:
// WML client library

import com.ibm.analytics.ngp.repository._

// Helper libraries

import scalaj.http.{Http, HttpOptions}
import scala.util.{Success, Failure}
import java.util.Base64
import java.nio.charset.StandardCharsets
import play.api.libs.json._

Authenticate to Watson Machine Learning service on Bluemix.

**Action**: Put authentication information from your instance of Watson Machine Learning service here.</div>

In [60]:
val service_path = "https://ibm-watson-ml.stage1.mybluemix.net"
val user = "95348cec-a0a6-4a6d-bc1b-a61648a58a9e"
val password = "9190e93a-c7ef-48c7-b849-44166cb2950f"

**Tip**: service_path, user and password can be found on **Service Credentials** tab of service instance created in Bluemix.

In [61]:
val client = MLRepositoryClient(service_path)
client.authorize(user, password)

Success(())

Create model artifact (abstraction layer).

In [62]:
val model_artifact = MLRepositoryArtifact(model_rf, training_data, "WML Product Line Prediction Model")

**Tip**: The MLRepositoryArtifact method expects a trained model object, training data, and a model name. (It is this model name that is displayed by the Watson Machine Learning service).

### 4.1: Save pipeline and model

In this subsection you will learn how to save pipeline and model artifacts to your Watson Machine Learning instance.

In [73]:
val saved_model = client.models.save(model_artifact).get

Get saved model metadata from Watson Machine Learning.

**Tip**: Use *meta.availableProps* to get the list of available props.

In [78]:
saved_model.meta.availableProps

Vector(trainingDataSchema, lastUpdated, label, inputDataSchema, modelVersionHref, modelType, version, modelHref, pipelineVersionHref, runtime, creationTime)

In [81]:
println("modelType: " + saved_model.meta.prop("modelType"))
println("trainingDataSchema: " + saved_model.meta.prop("trainingDataSchema"))
println("creationTime: " + saved_model.meta.prop("creationTime"))
println("modelVersionHref: " + saved_model.meta.prop("modelVersionHref"))
println("label: " + saved_model.meta.prop("label"))

modelType: Some(sparkml-model-2.0)
trainingDataSchema: Some({"type":"struct","fields":[{"name":"PRODUCT_LINE","type":"string","nullable":true,"metadata":{}},{"name":"GENDER","type":"string","nullable":true,"metadata":{}},{"name":"AGE","type":"integer","nullable":true,"metadata":{}},{"name":"MARITAL_STATUS","type":"string","nullable":true,"metadata":{}},{"name":"PROFESSION","type":"string","nullable":true,"metadata":{}}]})
creationTime: Some(2017-03-29T09:48:40.215Z)
modelVersionHref: Some(https://ibm-watson-ml.stage1.mybluemix.net/v2/artifacts/models/9c41d5a8-deb6-4c32-b169-1654ccc7a7b1/versions/a563ada8-d9be-4d12-9be5-9534d50c25de)
label: Some(PRODUCT_LINE)


**Tip**: **modelVersionHref** is our model unique indentifier in the Watson Machine Learning repository.

### 4.2: Load model and make predictions

In this subsection you will learn how to load back saved model from specified instance of Watson Machine Learning.

In [122]:
val model_version_href = saved_model.meta.prop("modelVersionHref").get
val loaded_model_artifact = client.models.version(model_version_href).get

You can print for example model name to make sure that model artifact has been loaded correctly.

In [129]:
loaded_model_artifact.name.mkString

WML Product Line Prediction Model

As you can see the name is correct. 

In [20]:
loaded_model_artifact match {
        case SparkPipelineModelLoader(Success(model)) => {
          val predictions = model.transform(prediction_data)
        }
        case SparkPipelineModelLoader(Failure(e)) => "Loading failed."
        case _ => println(s"Unexpected artifact class: ${loaded_model_artifact.getClass}")
    }
predictions.select("GENDER", "AGE", "MARITAL_STATUS", "PROFESSION","predictedLabel").show()

+------+---+--------------+-----------+--------------------+
|GENDER|AGE|MARITAL_STATUS| PROFESSION|      predictedLabel|
+------+---+--------------+-----------+--------------------+
|     F| 18|        Single|      Other|Personal Accessories|
|     F| 18|        Single|     Retail|Personal Accessories|
|     F| 19|        Single|Hospitality|   Camping Equipment|
|     F| 19|        Single|Hospitality|   Camping Equipment|
|     F| 19|        Single|Hospitality|   Camping Equipment|
|     F| 19|        Single|Hospitality|   Camping Equipment|
|     F| 19|        Single|      Other|Personal Accessories|
|     F| 19|        Single|      Other|Personal Accessories|
|     F| 19|        Single|      Other|Personal Accessories|
|     F| 19|        Single|      Other|Personal Accessories|
|     F| 19|        Single|      Other|Personal Accessories|
|     F| 19|        Single|      Other|Personal Accessories|
|     F| 19|        Single|      Other|Personal Accessories|
|     F| 19|        Sing

By tabulating a count, you can see which product line is the most popular.

In [21]:
predictions.select("predictedLabel").groupBy("predictedLabel").count().show()

+--------------------+-----+
|      predictedLabel|count|
+--------------------+-----+
|   Camping Equipment| 7238|
|      Golf Equipment|  677|
|Mountaineering Eq...|  192|
|Personal Accessories| 2753|
+--------------------+-----+



You have already learned how save and load the model from Watson Machine Learning repository.

<a id="scoring"></a>
## 5. Deploy and score in a Cloud

In this section you will learn how to create online scoring and to score a new data record by using the Watson Machine Learning REST API. 
For more information about REST APIs, see the [Swagger Documentation](http://watson-ml-api.mybluemix.net/).

To work with the Watson Machine Leraning REST API you must generate an access token. To do that you can use the following sample code:

In [65]:
// Get WML service instance token

val wml_auth_header = "Basic " + Base64.getEncoder.encodeToString((user + ":" + password).getBytes(StandardCharsets.UTF_8))
val wml_url = service_path + "/v2/identity/token"
val wml_response = Http(wml_url).header("Authorization", wml_auth_header).asString
val wmltoken_json: JsValue = Json.parse(wml_response.body)

val wmltoken = (wmltoken_json \ "token").asOpt[String] match {
    case Some(x) => x
    case None => ""
}

In [66]:
wmltoken

eyJhbGciOiJSUzUxMiIsInR5cCI6IkpXVCJ9.eyJ0ZW5hbnRJZCI6IjI4MjVhNjc3LTdiOTAtNGRjMi1iMjdiLTE2MWRlYjVlZjg5ZiIsImluc3RhbmNlSWQiOiIyODI1YTY3Ny03YjkwLTRkYzItYjI3Yi0xNjFkZWI1ZWY4OWYiLCJwbGFuSWQiOiIzZjZhY2Y0My1lZGU4LTQxM2EtYWM2OS1mOGFmM2JiMGNiZmUiLCJyZWdpb24iOiJ1cy1zb3V0aCIsInVzZXJJZCI6Ijk1MzQ4Y2VjLWEwYTYtNGE2ZC1iYzFiLWE2MTY0OGE1OGE5ZSIsImlzcyI6Imh0dHA6Ly8xMjkuNDEuMjI5LjE4ODo4MDgwL3YyL2lkZW50aXR5IiwiaWF0IjoxNDkwNzkwNDczLCJleHAiOjE0OTA4MTkyNzN9.GNhEj2tVI5EjMvL6WgL0N8t0lt7PbhHKv4DuMIiS0wxkWl5-1fD398cQSBZP3mg7KgNA4ZNh1ewUdTy1cQjdOqTYAJv--hevGmijtQBHt2b3GGxyo85raG-V0rVlBAwf1lil834kRoEKyKuufR7OAV-dhjdMLOZAXYvnPQuLSBLm6ZObPZNUgcb7EkAdEbQTYq66m5uGiHLWNZ_BR9JcUf8-QVyQQkH7VR7u3R6UjGGIKmxFkvGR7iwD9CIjDUXKv2tDoiB8ei14fjCYZP9gfN02kDGWy5JzbdikZqGoyD9Zkbg76Uvf4tqQpJloMsWvMyBBK8OK5kIfxadJb2nAWw

### 5.1: Create online scoring endpoint

Now you can create an online scoring endpoint. Execute the following sample code that uses the **modelVersionHref** value to create the scoring endpoint to the Bluemix repository.

In [70]:
val endpoint_online = "https://ibm-watson-ml.stage1.mybluemix.net/v2/online/deployments/"

In [71]:
val payload_artifactVersionHref = loaded_model_artifact.meta.prop("modelVersionHref").get
val payload_name = "Online scoring"
val payload_data_online = Json.stringify(Json.toJson(Map("artifactVersionHref" -> payload_artifactVersionHref, "name" -> payload_name)))

In [72]:
val response_online = Http(endpoint_online).postData(payload_data_online).header("Content-Type", "application/json").header("Authorization", wmltoken).option(HttpOptions.connTimeout(10000)).option(HttpOptions.readTimeout(50000)).asString

In [73]:
val scoring_href_json: JsValue = Json.parse(response_online.body)
val scoring_href = (scoring_href_json \ "entity" \ "scoringHref").asOpt[String] match {
    case Some(x) => x
    case None => ""
}

In [74]:
print(scoring_href)

https://ibm-watson-ml.stage1.mybluemix.net/32768/v2/scoring/1276

In [90]:
val payload_scoring = Json.stringify(Json.toJson(Map("record" -> Json.toJson(List(Json.toJson("M"), Json.toJson(55), Json.toJson("Single"), Json.toJson("Executive"))))))

In [91]:
payload_scoring

{"record":["M",55,"Single","Executive"]}

Now, you can send (PUT) new scoring records (new data) for which you would like to get predictions. To do that, execute the following sample code: 

In [92]:
val response_scoring = Http(scoring_href).postData(payload_scoring).header("Content-Type", "application/json").header("Authorization", "Bearer " + wmltoken).option(HttpOptions.method("PUT")).option(HttpOptions.connTimeout(10000)).option(HttpOptions.readTimeout(50000)).asString

In [93]:
print(response_scoring)

HttpResponse({"result":{"PROFESSION_IX":3.0,"GENDER_IX":0.0,"MARITAL_STATUS_IX":1.0,"GENDER":"M","features":{"values":[0.0,55.0,1.0,3.0]},"predictedLabel":"Mountaineering Equipment","prediction":2.0,"AGE":55,"MARITAL_STATUS":"Single","rawPrediction":{"values":[2.280235369131787,2.351474324815833,2.8897028385330197,2.283936941791034,0.19465052572832506]},"PROFESSION":"Executive","probability":{"values":[0.22802353691317875,0.23514743248158335,0.28897028385330203,0.2283936941791034,0.01946505257283251]}}},200,Map(Access-Control-Allow-Methods -> Vector(POST, GET, OPTIONS, PUT, DELETE), Access-Control-Allow-Origin -> Vector(*), Connection -> Vector(Keep-Alive), Content-Type -> Vector(application/json), Date -> Vector(Wed, 29 Mar 2017 12:56:43 GMT), Server -> Vector(nginx/1.11.5), Status -> Vector(HTTP/1.1 200 OK), Transfer-Encoding -> Vector(chunked), X-Backside-Transport -> Vector(OK OK), X-Global-Transaction-ID -> Vector(1517634271)))

As we can see we predict that a 55-year-old single male executive is interested in Camping Equipment (predictedLabel: Camping Equipment, prediction: 0.0).

<a id="summary"></a>
## 6. Summary and next steps 

You successfully completed this notebook! You learned how to use Apache Spark machine learning as well as Watson Machine Learning for model creation and deployment. Check out our [Online Documentation](http://datascience.ibm.com/docs/content/getting-started/get-started.html) for more samples, tutorials, documentation, how-tos, and blog posts. 
 
### Authors

**Umit Mert Cakmak** is Data Scientist in IBM with a track record of developing enterprise-level applications that substantially increases clients' ability to turn data into actionable insights.

Copyright © 2017 IBM. This notebook and its source code are released under the terms of the MIT License.